File: openreview/venue/process/proposed_assignment_pre_process.py
def process(client, edge, invitation):

    committee_name  = invitation.content['committee_name']['value']
    domain = client.get_group(edge.domain)
    custom_max_papers_id = domain.get_content_value(f'{committee_name.lower()}_custom_max_papers_id')
    custom_max_papers_default_value = None

    print(edge.id)

    if edge.ddate:
        return

    ## avoid validation during update
    if edge.tcdate != edge.tmdate:
        return

    custom_max_papers_invitation = openreview.tools.get_invitation(client, custom_max_papers_id) if custom_max_papers_id else None
    if custom_max_papers_invitation:
        custom_max_papers_default_value = custom_max_papers_invitation.edit['weight']['param'].get('default')

    ## Get quota
    edges=client.get_edges(invitation=custom_max_papers_id, tail=edge.tail)

    custom_max_papers=edges[0].weight if edges else custom_max_papers_default_value

    if not custom_max_papers:
        return edge

    assignment_edges=client.get_edges(invitation=edge.invitation, label=edge.label, tail=edge.tail)

    if len(assignment_edges) >= custom_max_papers:
        profile=openreview.tools.get_profile(client, edge.tail)
        raise openreview.OpenReviewException(f'Max Papers allowed reached for {profile.get_preferred_name(pretty=True) if profile else edge.tail}')

    return edge

File: openreview/venue/process/proposed_assignment_pre_process.js
async function process(client, edge, invitation) {
  client.throwErrors = true

  const committeeName = invitation.content.committee_name?.value;
  const { groups } = await client.getGroups({ id: invitation.domain });
  const domain = groups[0];
  const reviewersName = domain.content['reviewers_name']?.value
  const areaChairsName = domain.content['area_chairs_name']?.value
  const roleMap = {
    [reviewersName]: 'reviewers',
    [areaChairsName]: 'area_chairs',
  };
  const internalRole = roleMap[committeeName]
  const customMaxPapersId = domain.content[`${internalRole}_custom_max_papers_id`]?.value;
  const quota = domain.content?.[`submission_assignment_max_${internalRole}`]?.value

  if (edge.ddate) {
    return
  }

  if (edge.tcdate !== edge.tmdate) {
    return
  }

  if (!customMaxPapersId) {
    return
  }

  const [res1, res2, res3] = await Promise.all([
    client.getEdges({ invitation: customMaxPapersId, tail: edge.tail }),
    client.getEdges({ invitation: edge.invitation, label: edge.label, tail: edge.tail }),
    client.getEdges({ invitation: edge.invitation, label: edge.label, head: edge.head })
  ])
  let customMaxPapers = (res1.count > 0 && res1.edges[0].weight) || 0;
  const assignmentEdges = res2.edges;
  const submissionEdges = res3.edges;

  if (!customMaxPapers) {
    const { invitations } = await client.getInvitations({ id: customMaxPapersId });
    customMaxPapers = invitations[0].edge.weight.param.default;
  }

  if (assignmentEdges.length >= customMaxPapers) {
    const { profiles } = await client.getProfiles({ id: edge.tail });
    const profile = profiles[0];
    return Promise.reject(new OpenReviewError({ name: 'Error', message: `Max Papers allowed reached for ${Tools.getPreferredName(profile)}` }));
  }

  if (quota && (submissionEdges.length + 1) > quota) {
    return Promise.reject(new OpenReviewError({ name: 'Error', message: `You cannot assign more than ${quota} ${committeeName.toLowerCase().replaceAll('_',' ')} to this paper` }));
  }

}
